enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,335 @@
+   FROM registry.ci.openshift.org/openshift/release:golang-1.15 AS builder
+   ... build the image ...
+
+   FROM ose-aws-efs-utils-base
+   COPY --from=builder <the driver binary> /usr/bin/
+   ```
+
+It would be nice if we did not ship `ose-aws-efs-utils-base` anywhere, so people can consume the EFS utils only through
+the EFS operator. **We do not want to maintain it for everyone.**
+
+### Code organization
+
+* Reuse existing `github.com/openshift/aws-efs-operator` for the operator.
+  * Move the controller-runtime based operator one to a separate branch (`release-4.7`, even though it contains all
+    previous releases?)
+  * Continue with library-go based one in `master`.
+* Fork the CSI driver into `github.com/openshift/aws-efs-csi-driver`.
+* Fork `github.com/aws/efs-utils` into `github.com/openshift/aws-efs-utils`.
+
+### Open Questions
+
+* How to graduate community operator into a supported one?
+  * The community one does not go through registry.redhat.io.
+    * The "community" team is willing to do anything that's necessary to help with the transition.
+  * The community one has only a single channel, `stable`.
+  * We want the community one working in 4.7 and older and the supported one in 4.8 and newer.
+  * It should be possible to ship new community updates in 4.7 channel/branch to fix potential bugs there and/or
+    make the transition as seamless as possible.
+
+### Test Plan
+
+* Automatic jobs:
+  * Run CSI certification tests with the CSI driver:
+     1. Install OCP on AWS.
+     2. Install the operator.
+     3. Create AWS EFS share (this may require extra privileges in CI and figure out the networking!)
+     4. Set up dynamic provisioning in the driver (by creating a StorageClass). This assumes dynamic provisioning
+        is implemented (released) upstream before we start testing.
+     5. Run CSI certification tests.
+  * Run upgrade tests:
+     1. Install OCP on AWS.
+     2. Install the operator.
+     3. Create AWS EFS share (this may require extra privileges in CI and figure out the networking!)
+     4. Create an dummy app that uses EFS share.
+     5. Upgrade the EFS operator (via OLM).
+     6. Check the app survived.
+
+* Manual tests:
+  * Upgrade from the community CSI driver. With some application pods using volumes provided by the community driver.
+    There should be no (long) disruption of the application.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+* The functionality described above is implemented.
+* At least manual testing passed (CI may need dynamic provisioning in the driver).
+* End user documentation exists.
+
+#### Tech Preview -> GA
+
+TODO. Mainly all tests in CI & feedback from the users (OSD!)
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+- Document SLOs for the component
+- Conduct load testing
+
+#### Removing a deprecated feature
+
+TODO: shall we deprecate `SharedVolume`?
```

Comment:
Probably. You convinced me that it carries a security risk that we can't count on users understanding.

...and if we're going to do this, doing it on the boundary of community=>supported would probably make the most sense. Happy to discuss strategy here.

Reply:
I'd love to have the functionality provided by `SharedVolume` implemented in the CSI driver itself. Maybe we can focus on that rather than supporting this CR?

Reply:
> I'd love to have the functionality provided by `SharedVolume` implemented in the CSI driver itself.

Can you explain this a bit further, please? Today all `SharedVolume` does is accept a FS and AP ID and create a PV/PVC pair for it. It's basically just a shorthand way to create dozens of lines worth of resource definition with two fields. And it's also an RBAC back door -- which was deliberate, but isn't worth the security in my (new and improved) opinion.

Reply:
I meant the CSI driver should support dynamic provisioning, which has been implemented recently (as you pointed out above).